modules/transaction-payment/src/lib.rs

```diff
@@ -824,12 +825,15 @@ where
 					TokenExchangeRate::<T>::contains_key(currency_id),
 					Error::<T>::InvalidToken
 				);
-				let fee_amount = if T::DefaultFeeTokens::get().contains(currency_id) {
-					fee.saturating_add(T::AlternativeFeeSurplus::get().mul_ceil(fee))
+				let fee = Self::check_native_is_not_enough(who, fee, reason).map_or_else(|| fee, |amount| amount);
+				let alternative_fee_surplus = T::AlternativeFeeSurplus::get().mul_ceil(fee);
+				let custom_fee_surplus = T::CustomFeeSurplus::get().mul_ceil(fee);
+				let (fee_amount, fee_surplus) = if T::DefaultFeeTokens::get().contains(currency_id) {
+					(fee.saturating_add(alternative_fee_surplus), alternative_fee_surplus)
 				} else {
-					custom_fee_amount
+					(fee.saturating_add(custom_fee_surplus), custom_fee_surplus)
 				};
-				Self::swap_from_pool_or_dex(who, fee_amount, *currency_id).map(|_| (who.clone(), fee_amount))
```

Comment:
this parameter is wrong, should passing surplus, not fee+surplus.

Also need consider ED when user has none native token.